File: trunk/NowPlaying/Android/src/org/metasyntactic/caches/scores/GoogleScoreProvider.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.caches.scores;

import com.google.protobuf.InvalidProtocolBufferException;
import org.metasyntactic.NowPlayingApplication;
import org.metasyntactic.NowPlayingModel;
import org.metasyntactic.caches.UserLocationCache;
import org.metasyntactic.data.Location;
import org.metasyntactic.data.Score;
import org.metasyntactic.protobuf.NowPlaying;
import org.metasyntactic.time.Days;
import org.metasyntactic.utilities.ExceptionUtilities;
import org.metasyntactic.utilities.NetworkUtilities;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;

import static java.lang.Math.max;
import static java.lang.Math.min;

import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class GoogleScoreProvider extends AbstractScoreProvider {
  public GoogleScoreProvider(final NowPlayingModel model) {
    super(model);
  }

  @Override protected String getProviderName() {
    return "Google";
  }

  private String getUrl() {
    final Location location = UserLocationCache.downloadUserAddressLocationBackgroundEntryPoint(
        getModel().getUserAddress());

    if (location == null || isNullOrEmpty(location.getPostalCode())) {
      return null;
    }

    final String country = isNullOrEmpty(
        location.getCountry()) ? Locale.getDefault().getCountry() : location.getCountry();

    //Debug.startMethodTracing("getUrlDaysBetween", 1 << 24);
    int days = Days.daysBetween(new Date(), getModel().getSearchDate());
    //Debug.stopMethodTracing();
    days = min(max(days, 0), 7);

    return "http://" + NowPlayingApplication.host + ".appspot.com/LookupTheaterListings2?country=" + country + "&language=" + Locale.getDefault().getLanguage() + "&day=" + days + "&format=pb" + "&latitude=" + (int) (location.getLatitude() * 1000000) + "&longitude=" + (int) (location.getLongitude() * 1000000);
  }

  @Override protected String lookupServerHash() {
    String address = getUrl();
    if (isNullOrEmpty(address)) {
      return address;
    }
    address += "&hash=true";
    return NetworkUtilities.downloadString(address, true);
  }

  @Override protected Map<String, Score> lookupServerScores() {
    final String address = getUrl();
    if (isNullOrEmpty(address)) {
      return Collections.emptyMap();
    }
    final byte[] data = NetworkUtilities.download(address, true);

    if (data != null) {
      final NowPlaying.TheaterListingsProto theaterListings;
      try {
        theaterListings = NowPlaying.TheaterListingsProto.parseFrom(data);
      } catch (final InvalidProtocolBufferException e) {
        ExceptionUtilities.log(GoogleScoreProvider.class, "lookupServerRatings", e);
        return null;
      }

      final Map<String, Score> ratings = new HashMap<String, Score>();

      for (final NowPlaying.MovieProto movieProto : theaterListings.getMoviesList()) {
        final String identifier = movieProto.getIdentifier();
        final String title = movieProto.getTitle();
        int value = -1;
        if (movieProto.hasScore()) {
          value = movieProto.getScore();
        }

        final Score score = new Score(title, "", String.valueOf(value), "google", identifier);
        ratings.put(score.getCanonicalTitle(), score);
      }

      return ratings;
    }

    return null;
  }
}

File: NowPlaying/iPhone/Classes/GoogleScoreProvider.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "GoogleScoreProvider.h"

#import "Application.h"
#import "DateUtilities.h"
#import "LocaleUtilities.h"
#import "Location.h"
#import "Model.h"
#import "NetworkUtilities.h"
#import "NowPlaying.pb.h"
#import "Score.h"
#import "ScoreCache.h"
#import "StringUtilities.h"
#import "UserLocationCache.h"
#import "Utilities.h"
#import "XmlElement.h"


@implementation GoogleScoreProvider

- (void) dealloc {
    [super dealloc];
}


+ (GoogleScoreProvider*) providerWithModel:(Model*) model {
    return [[[GoogleScoreProvider alloc] initWithModel:model] autorelease];
}


- (NSString*) providerName {
    return @"Google";
}


- (NSString*) serverUrl {
    Location* location = [self.model.userLocationCache locationForUserAddress:self.model.userAddress];

    if (location.postalCode == nil) {
        return nil;
    }

    NSString* country = location.country.length == 0 ? [LocaleUtilities isoCountry]
    : location.country;


    NSDateComponents* components = [[NSCalendar currentCalendar] components:NSDayCalendarUnit
                                                                   fromDate:[DateUtilities today]
                                                                     toDate:self.model.searchDate
                                                                    options:0];
    NSInteger day = components.day;
    day = MIN(MAX(day, 0), 7);

    NSString* address = [NSString stringWithFormat:
                         @"http://%@.appspot.com/LookupTheaterListings2?country=%@&language=%@&postalcode=%@&day=%d&format=pb&latitude=%d&longitude=%d",
                         [Application host],
                         country,
                         [LocaleUtilities isoLanguage],
                         [StringUtilities stringByAddingPercentEscapes:location.postalCode],
                         day,
                         (int)(location.latitude * 1000000),
                         (int)(location.longitude * 1000000)];

    return address;
}


- (NSString*) lookupServerHash {
    NSString* baseAddress = [self serverUrl];
    NSString* address = [baseAddress stringByAppendingString:@"&hash=true"];
    NSString* value = [NetworkUtilities stringWithContentsOfAddress:address];
    return value;
}


- (NSDictionary*) lookupServerScores {
    NSString* address = [self serverUrl];
    NSData* data = [NetworkUtilities dataWithContentsOfAddress:address];
    if (data != nil) {
        @try {
            TheaterListingsProto* theaterListings = [TheaterListingsProto parseFromData:data];
            NSArray* movieProtos = theaterListings.moviesList;

            NSMutableDictionary* ratings = [NSMutableDictionary dictionary];

            for (MovieProto* movieProto in movieProtos) {
                NSString* identifier = movieProto.identifier;
                NSString* title = movieProto.title;
                NSInteger score = -1;
                if (movieProto.hasScore) {
                    score = movieProto.score;
                }

                Score* info = [Score scoreWithTitle:title
                                           synopsis:@""
                                              score:[NSString stringWithFormat:@"%d", score]
                                           provider:@"google"
                                         identifier:identifier];

                [ratings setObject:info forKey:info.canonicalTitle];
            }

            return ratings;
        } @catch (NSException* e) {
        }
    }

    return nil;
}

@end